File: witterungsstation_py/model/api_client.py
from __future__ import annotations

import json
from dataclasses import dataclass
from typing import Any, Optional

import requests

from model.api_paths import path_variants


@dataclass
class ApiError(Exception):
    status: int
    method: str
    path: str
    data: Any

    def __str__(self) -> str:
        return f"HTTP {self.status} {self.method} {self.path}".strip()


class ApiClient:
    def __init__(self, base_url: str, session: Optional[requests.Session] = None) -> None:
        self.base_url = base_url.rstrip("/")
        self._session = session or requests.Session()
        self._session.headers.update({"Content-Type": "application/json"})

    @staticmethod
    def _parse_payload(response: requests.Response) -> Any:
        text = response.text or ""
        if not text:
            return {}
        try:
            return json.loads(text)
        except Exception:
            return {"raw": text}

    def request(
        self,
        method: str,
        path: str,
        *,
        json_body: Any = None,
        timeout: float = 5.0,
    ) -> Any:
        m = str(method or "GET").upper()
        last_error: Exception | None = None

        for variant in path_variants(path):
            url = f"{self.base_url}{variant}"
            try:
                response = self._session.request(m, url, json=json_body, timeout=timeout)
                payload = self._parse_payload(response)
                if not response.ok:
                    raise ApiError(
                        status=int(response.status_code),
                        method=m,
                        path=variant,
                        data=payload,
                    )
                return payload
            except ApiError as error:
                last_error = error
                if error.status != 404:
                    raise
            except Exception as error:
                last_error = error
                raise

        if last_error is not None:
            raise last_error
        raise RuntimeError("Unknown API error")

    def get(self, path: str, *, timeout: float = 5.0) -> Any:
        return self.request("GET", path, timeout=timeout)

    def post(self, path: str, *, json_body: Any, timeout: float = 5.0) -> Any:
        return self.request("POST", path, json_body=json_body, timeout=timeout)

File: frontend/src/services/apiClient.ts
import { createApiError } from './apiErrors'
import { getPathVariants } from './apiPath'

export interface RequestOptions {
  body?: unknown
  token?: string
  form?: boolean
}

export class ApiClient {
  baseUrl: string
  onUnauthorized: ((payload: { status: number; method: string; path: string; data: unknown }) => void) | null

  constructor(baseUrl: string, { onUnauthorized }: { onUnauthorized?: (payload: { status: number; method: string; path: string; data: unknown }) => void } = {}) {
    this.baseUrl = baseUrl
    this.onUnauthorized = typeof onUnauthorized === 'function' ? onUnauthorized : null
  }

  async request(method: string, path: string, options: RequestOptions = {}): Promise<unknown> {
    const { body, token, form = false } = options
    const headers: Record<string, string> = {}
    let payload: string | undefined
    const hasAuthToken = Boolean(String(token || '').trim())

    if (body !== undefined) {
      if (form) {
        headers['Content-Type'] = 'application/x-www-form-urlencoded'
        payload = new URLSearchParams(body as Record<string, string>).toString()
      } else {
        headers['Content-Type'] = 'application/json'
        payload = JSON.stringify(body)
      }
    }

    if (token) {
      headers.Authorization = `Bearer ${token}`
    }

    const tryPaths = getPathVariants(path)

    let lastError: unknown = null
    for (const p of tryPaths) {
      const url = new URL(p, this.baseUrl).toString()
      try {
        const res = await fetch(url, { method, headers, body: payload })
        const text = await res.text()
        let data: unknown = {}
        if (text) {
          try {
            data = JSON.parse(text)
          } catch {
            data = { raw: text }
          }
        }

        if (!res.ok) {
          if (res.status === 401 && hasAuthToken && this.onUnauthorized) {
            try {
              this.onUnauthorized({ status: res.status, method, path: p, data })
            } catch {
              // Ignore unauthorized callback failures.
            }
          }

          throw createApiError({
            status: res.status,
            method,
            path: p,
            data,
          })
        }

        return data
      } catch (error) {
        lastError = error
        const status = Number((error as { status?: number })?.status) || 0
        if (status !== 404) {
          break
        }
      }
    }

    throw lastError || new Error('Unknown API error')
  }

  get(path: string, options: Omit<RequestOptions, 'body' | 'form'> = {}): Promise<unknown> {
    return this.request('GET', path, { ...options, body: undefined, form: undefined })
  }

  post(path: string, body: unknown, options: Omit<RequestOptions, 'body'> = {}): Promise<unknown> {
    return this.request('POST', path, { ...options, body })
  }

  postForm(path: string, body: Record<string, unknown>, options: Omit<RequestOptions, 'body' | 'form'> = {}): Promise<unknown> {
    return this.request('POST', path, { ...options, body, form: true })
  }

  put(path: string, body: unknown, options: Omit<RequestOptions, 'body'> = {}): Promise<unknown> {
    return this.request('PUT', path, { ...options, body })
  }

  delete(path: string, options: Omit<RequestOptions, 'body' | 'form'> = {}): Promise<unknown> {
    return this.request('DELETE', path, { ...options, body: undefined, form: undefined })
  }
}
